extractor/src/test/java/org/schabi/newpipe/extractor/services/youtube/YoutubeDashManifestCreatorTest.java

```diff
@@ -0,0 +1,620 @@
+package org.schabi.newpipe.extractor.services.youtube;
+
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
+import org.schabi.newpipe.downloader.DownloaderTestImpl;
+import org.schabi.newpipe.extractor.NewPipe;
+import org.schabi.newpipe.extractor.services.youtube.extractors.YoutubeStreamExtractor;
+import org.schabi.newpipe.extractor.stream.DeliveryMethod;
+import org.schabi.newpipe.extractor.stream.Stream;
+import org.w3c.dom.Document;
+import org.w3c.dom.Element;
+import org.w3c.dom.NodeList;
+import org.xml.sax.InputSource;
+
+import javax.annotation.Nonnull;
+import javax.xml.parsers.DocumentBuilder;
+import javax.xml.parsers.DocumentBuilderFactory;
+import java.io.StringReader;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.List;
+import java.util.Random;
+
+import static org.junit.jupiter.api.Assertions.*;
+import static org.schabi.newpipe.extractor.ServiceList.YouTube;
+import static org.schabi.newpipe.extractor.utils.Utils.isBlank;
+
+/**
+ * Test for {@link YoutubeDashManifestCreator}.
+ *
+ * <p>
+ * We cannot test the generation of DASH manifests for ended livestreams because these videos will
+ * be re-encoded as normal videos later, so we can't use a specific video.
+ * </p>
+ *
+ * <p>
+ * The generation of DASH manifests for OTF streams, which can be tested, uses a video licenced
+ * under the Creative Commons Attribution licence (reuse allowed):
+ * {@code https://www.youtube.com/watch?v=DJ8GQUNUXGM}
+ * </p>
+ *
+ * <p>
+ * We couldn't use mocks for these tests because the streaming URLs needs to fetched and the IP
+ * address used to get these URLs is required (used as a param in the URLs; without it, video
+ * servers return 403/Forbidden HTTP response code).
+ * </p>
+ *
+ * <p>
+ * So the real downloader will be used everytime on this test class.
+ * </p>
+ */
+class YoutubeDashManifestCreatorTest {
+    // Setting a higher number may let Google video servers return a lot of 403s
+    private static final int MAXIMUM_NUMBER_OF_STREAMS_TO_TEST = 3;
+
+    public static class testGenerationOfOtfAndProgressiveManifests {
```

Comment:
Could you develop a bit what do you mean? I think I understood, but I want to be sure.

Reply:
Well my problem here is that you have one (maybe 2-3) actual tests with a ton of conditions.

Problems that arise of this:
1. If one condition/assertion fails no other assertions get tested (that's why there is also e.g. an assertAll)
2. The tests are complex and not easily maintainable/understandable, we might miss something important iniside a test and don't notice it
3. Parts of the test are not stateless - Maybe if the tests just work because a random cache got not reset. Example: https://github.com/TeamNewPipe/NewPipeExtractor/pull/810/files/5e3740bc921150574e09e564f75c34927929c2b1#diff-4b831cc5e11b2dc2a96019af120a199acd27396bf5230d59cdfc86a24c882046R90-R102

As written inside https://softwareengineering.stackexchange.com/a/171502 I highly recommend the "Arrange, Act, Assert"/"GivenWhenThen"-pattern.

Reply:
Test has been improved a lot by Stypox. Let me know in another review/comment if it is good for you or not.